source/common/network/address_impl.h

```diff
@@ -245,6 +250,32 @@ class PipeInstance : public InstanceBase {
   PipeHelper pipe_;
 };
 
+class EnvoyInternalInstance : public InstanceBase {
+public:
+  /**
+   * Construct from a string name.
+   */
+  explicit EnvoyInternalInstance(const std::string& envoy_listener_name,
+                                 const SocketInterface* sock_interface = nullptr);
+
+  // Network::Address::Instance
+  bool operator==(const Instance& rhs) const override;
+  const Ip* ip() const override { return nullptr; }
+  const Pipe* pipe() const override { return nullptr; }
+  const EnvoyInternalAddress* envoyInternalAddress() const override { return &internal_address_; }
+  const sockaddr* sockAddr() const override { return nullptr; }
```

Comment:
Can we have a TODO for auditing call sites and making sure they (and transitive calls) are safe?  I see at least one sockaddr() call and as discussed the other one isn't actually failproof yet.